File: go-ms/domain/usecase/cases_use_case.go
package usecase

import (
	hello "maocq/go-ms/domain/model/hello/gateways"
	"maocq/go-ms/domain/model/primes"
)

type CasesUseCase struct {
	HelloRepository hello.HelloRepository
}

func (c *CasesUseCase) CaseOne(latency string) (string, error) {
	if _, err := c.HelloRepository.Hello(latency); err != nil {
		return "", err
	}
	result := primes.Primes(1000)

	return result, nil
}

func (c *CasesUseCase) CaseTwo(latency string) (string, error) {
	return c.HelloRepository.Hello(latency)
}

func (c *CasesUseCase) CaseThree(n int) string {
	return primes.Primes(n)
}

File: java-spring3-no-native-ms/src/main/java/com/maocq/spring/domain/usecase/cases/CasesUseCase.java
package com.maocq.spring.domain.usecase.cases;

import com.maocq.spring.Application;
import com.maocq.spring.domain.model.hello.gateways.HelloGateway;
import com.maocq.spring.domain.model.primes.Primes;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

@Service
public class CasesUseCase {
    private final HelloGateway helloGateway;

    public CasesUseCase(HelloGateway helloGateway) {
        this.helloGateway = helloGateway;
    }

    public Mono<String> caseOne(int latency) {
        return helloGateway.hello(latency)
                .flatMap(x -> Mono.just(Primes.primes(1000)));
    }

    public Mono<String> caseTwo(int latency) {
        return helloGateway.hello(latency);
    }

    public Mono<String> caseThree(int n) {
        return Mono.defer(() -> Mono.just(Primes.primes(n)))
                .subscribeOn(Schedulers.fromExecutor(Application.cpuExecutor));
    }
}
